src/sentry/api/endpoints/user_authenticator_enroll.py

```diff
@@ -0,0 +1,195 @@
+        """
+
+        interface = Authenticator.objects.get_interface(user, interface_id)
+
+        # Not all interfaces allow multi enrollment
+        if interface.is_enrolled and not interface.allow_multi_enrollment:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        # User is not enrolled in auth interface:
+        # - display configuration form
+        response = serialize(interface)
+        response['form'] = get_serializer_field_metadata(
+            serializer_map[interface_id]
+        )
+
+        # U2fInterface has no 'secret' attribute
+        try:
+            response['secret'] = interface.secret
+        except AttributeError:
+            pass
+
+        if interface_id == 'totp':
+            response['qrcode'] = interface.get_provision_qrcode(user.email)
+
+        if interface_id == 'u2f':
+            response['challenge'] = interface.start_enrollment()
+
+        return Response(response)
+
+    @sudo_required
+    def post(self, request, user, interface_id):
+        """
+        Enroll in authenticator interface
+        `````````````````````````````````
+
+        :pparam string user_id: user id or "me" for current user
+        :pparam string interface_id: interface id
+
+        :auth: required
+        """
+
+        # Using `request.user` here because doesn't seem like a superuser should be able to
+        # set a user's 2fa?
+
+        # start activation
+        serializer_cls = serializer_map.get(interface_id, None)
+
+        if serializer_cls is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        serializer = serializer_cls(data=request.DATA)
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        interface = Authenticator.objects.get_interface(request.user, interface_id)
+
+        # Not all interfaces allow multi enrollment
+        #
+        # This is probably un-needed because we catch
+        # `Authenticator.AlreadyEnrolled` when attempting to enroll
+        if interface.is_enrolled and not interface.allow_multi_enrollment:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            interface.secret = request.DATA['secret']
+        except KeyError:
+            pass
+
+        # Need to update interface with phone number before validating OTP
+        if 'phone' in request.DATA:
+            interface.phone_number = serializer.data['phone']
+
+            # Disregarding value of 'otp', if no OTP was provided,
+            # send text message to phone number with OTP
+            if 'otp' not in request.DATA:
+                if interface.send_text(for_enrollment=True, request=request._request):
+                    return Response(status=status.HTTP_204_NO_CONTENT)
+                else:
+                    # Error sending text message
+                    return Response(status=status.HTTP_400_BAD_REQUEST)
```

Comment:
I don't think it's reasonable that we're generically returning back `400 Bad Request` for errors without any context. In this case, a 400 doesn't even make sense. 400's are historically really used for malformed HTTP request, not based on just bad user input or errors.

But I really dislike the idea of returning a plain 400 with no body in lots of cases without any reasoning why.

Reply:
Provided we add a body w/ error reason, what would be a more appropriate http code to return?

Reply:
Well, here's an error. So a 500 is appropriate. Wouldn't this be a product of an error from Twilio?